[RTM] Sanitize the file names of uploaded files

leofeyer opened the pull request:
This PR will sanitize the file names of uploaded files and make sure that they are not exceeding the length of the database field.

leofeyer committed:
Sanitize the file names of uploaded files.

Toflar commented:
Looks good to me.

ausi commented:
Should we handle uploads from the frontend and backend the same way?

I still vote for a whitelist like `[^A-Za-z0-9_.-]`, at least for the frontend.

leofeyer commented:
> Should we handle uploads from the frontend and backend the same way?

Yes, why not? Wouldn't it be confusing if we didn't?

> I still vote for a whitelist like `[^A-Za-z0-9_.-]`, at least for the frontend.

You suggested `/(?:["\'<>]|[^\pL\pN\pP\pS])+/u`, right? That does not allow the "Mark" group (characters intended to be combined with another characters, e.g. accents, umlauts, enclosing boxes, etc.). Did you omit this on purpose?

Actually, I do have a general problem with the `utf8_romanize()` call. It seems that Contao as well as the browsers can handle UTF-8 characters, therefore I wonder why we are romanizing at all?

leofeyer commented:
Also, I'm not sure if we should forbid brackets and quotes by default. What about the following file name:

`reports "cleaned" < 2015.xls`

Isn't this a perfectly valid file name?

ausi commented:
> You suggested `/(?:["\'<>]|[^\pL\pN\pP\pS])+/u`, right?

Only if we really don’t want to use `[^A-Za-z0-9_.-]`.

> Actually, I do have a general problem with the `utf8_romanize()` call. It seems that Contao as well as the browsers can handle UTF-8 characters, therefore I wonder why we are romanizing at all?

IMO `utf8_romanize()` would only make sense if we use `[^A-Za-z0-9_.-]`.

>`reports "cleaned" < 2015.xls`
> Isn't this a perfectly valid file name?

It depends on the operating system and the file system. `<` and `"` are not allowed on Windows for example:

![windows-8-filename](https://cloud.githubusercontent.com/assets/367169/7937520/bdafa940-0940-11e5-9013-ae3a3c33567e.png)

IMO it’s perfectly fine to rename the file to `reports-cleaned-2015.xls` if it was uploaded via the frontend.

ausi commented:
Maybe it’s better to discuss this issue in a Mumble call?

leofeyer commented:
We'll have our dev call tomorrow at 4:30pm. Do you have time?

ausi commented:
Yes.

leofeyer committed:
Do not apply utf8_romanize() in the sanitizeFileName() method and adjust the reggae.

leofeyer committed:
Remove special characters not supported on e.g. Windows

leofeyer committed:
Fix the regular expression.

leofeyer committed:
Throw an exception if the file name cannot be sanitized.

leofeyer committed:
Catch the exception in the upload forms.

leofeyer committed:
Double encode the tool tip titles in the file manager.

Toflar commented on system/modules/core/classes/FileUpload.php:
```diff
@@ -103,13 +103,28 @@ public function uploadTo($strTarget)
+			}
+			catch (\InvalidArgumentException $e)
+			{
+				$this->log('File "'.$file['name'].'" cannot be sanitized' , __METHOD__, TL_ERROR);
```
So the end user does not get any message?

discordier replied:
IMO it should read:
```php
\Message::addError($GLOBALS['TL_LANG']['ERR']['invalid_filename']);
$this->log('File "'.$file['name'].'" cannot be sanitized' , __METHOD__, TL_ERROR);
$this->blnHasError = true;
```
With the corresponding language string added.

Toflar commented on system/modules/core/forms/FormFileUpload.php:
```diff
@@ -108,8 +108,27 @@ public function validate()
+		}
+		catch (\InvalidArgumentException $e)
+		{
+			$this->addError(sprintf($GLOBALS['TL_LANG']['ERR']['filename'], $file['name']));
```
Not sure if this is a good idea because it would output the invalid name which might contain anything bad!

discordier replied:
As I already said on mumble before, do NOT output the filename as it is broken and the resulting message is unpredictable.

leofeyer replied:
So do you want us not to tell the user which of their files is broken?

Toflar commented on system/modules/core/forms/FormFileUpload.php:
```diff
@@ -108,8 +108,27 @@ public function validate()
+		// Invalid file name
+		if (!\Validator::isValidFileName($file['name']))
+		{
+			$this->addError(sprintf($GLOBALS['TL_LANG']['ERR']['filename'], $file['name']));
```
Same here

Toflar commented on system/modules/core/library/Contao/Validator.php:
```diff
@@ -388,4 +388,35 @@ public static function isInsecurePath($strPath)
+		}
+
+		// Must not be longer than 255 characters
+		if (utf8_strlen($strName) > 255)
```
Imho, this should be the first check because it's likely the fastest of them all :)

ausi replied:
IMO the preg_match in the first `if` is faster than `utf8_strlen()`

discordier replied:
But it might choke on invalid UTF chars which are checked in 409.

ausi replied:
409 doesn’t check if it is valid UTF-8, in fact it can be bypassed with invalid UTF-8:

```php
preg_match('/[\pC]/u', "normal string") === 0
preg_match('/[\pC]/u', "string with null byte \0") === 1
preg_match('/[\pC]/u', "string with null byte \0 and invalid UTF-8 \xC3") === false
```

So we should change the check to `if (preg_match('/[\pC]/u', $strName) !== 0)`.

Toflar replied:
I think this test should be extended to

```php
// Must not be longer than 255 characters but at least 1 character
$length = utf8_strlen($strName);
if ($length > 255 || $length < 1)
```

Otherwise an empty string would be a valid file name.

ausi commented:
We should also escape [system/modules/core/templates/backend/be_popup.html5:52](https://github.com/contao/core/blob/8f678aaf08102432f943241a90d13c1c308741bd/system/modules/core/templates/backend/be_popup.html5#L52).

ausi commented on system/modules/core/dca/tl_files.php:
```diff
@@ -621,7 +621,7 @@ public function showFile($row, $href, $label, $title, $icon, $attributes)
 		else
 		{
-			return '<a href="contao/popup.php?src=' . base64_encode($row['id']) . '" title="'.specialchars($title).'"'.$attributes.' onclick="Backend.openModalIframe({\'width\':'.$row['popupWidth'].',\'title\':\''.str_replace("'", "\\'", $row['fileNameEncoded']).'\',\'url\':this.href,\'height\':'.$row['popupHeight'].'});return false">'.Image::getHtml($icon, $label).'</a> ';
+			return '<a href="contao/popup.php?src=' . base64_encode($row['id']) . '" title="'.htmlspecialchars($title, ENT_COMPAT, Config::get('characterSet')).'"'.$attributes.' onclick="Backend.openModalIframe({\'width\':'.$row['popupWidth'].',\'title\':\''.str_replace("'", "\\'", $row['fileNameEncoded']).'\',\'url\':this.href,\'height\':'.$row['popupHeight'].'});return false">'.Image::getHtml($icon, $la ...
```
`$row['fileNameEncoded']` should also get encoded in this line.

leofeyer replied:
`$row['fileNameEncoded']` is encoded already, isn't it?

leofeyer replied:
No, you are right. I have changed it accordingly.

leofeyer committed:
Fix some issues found during the review process.

leofeyer committed:
Add a flag to the specialchars() method and also encode the modal window title.

leofeyer committed:
Do not output the invalid file name in the error message.

leofeyer commented:
I have adjusted everything according to your comments.

leofeyer committed:
Do not escape the pipe character inside a character class.

In most regex flavors, the only special characters or meta characters inside a character class are the closing bracket (]), the backslash (\), the caret (^), and the hyphen (-).

leofeyer committed:
Return false for empty file names in Validator::isValidFileName().

leofeyer commented:
Merged in 6aaa26c5051b23b7881b4d73bd816aab567284d6.
